### ISSUE TITLE
error: static assertion failed: Protobuf only supports C++14 and newer

### ISSUE THREAD

**Issue by yurivict** [onnx/onnx#4992]
# Bug Report

### Describe the bug
The build fails because protobuf header is built with -std=gnu++11 when it requires c++17.
I added -std=c++20 to CXXFLAGS but this is added earlier than -std=gnu++11 and doesn't take effect.

```
[ 10%] Building CXX object CMakeFiles/onnx_proto.dir/onnx/onnx-ml.pb.cc.o
/depot/gcc-12.2.0/bin/g++ -DONNX_API="__attribute__((__visibility__(\"default\")))" -DONNX_ML=1 -DONNX_NAMESPACE=onnx -D__STDC_FORMAT_MACROS -I/u/yuriv/nn-trainer/deps/onnx -isystem /u/yuriv/local/include -std=c++20 -Wnon-virtual-dtor -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=gnu++11 -MD -MT CMakeFiles/onnx_proto.dir/onnx/onnx-ml.pb.cc.o -MF CMakeFiles/onnx_proto.dir/onnx/onnx-ml.pb.cc.o.d -o CMakeFiles/onnx_proto.dir/onnx/onnx-ml.pb.cc.o -c /u/yuriv/nn-trainer/deps/onnx/onnx/onnx-ml.pb.cc
In file included from /u/yuriv/nn-trainer/deps/onnx/onnx/onnx-ml.pb.h:11,
                 from /u/yuriv/nn-trainer/deps/onnx/onnx/onnx-ml.pb.cc:4:
/u/yuriv/local/include/google/protobuf/port_def.inc:200:15: error: static assertion failed: Protobuf only supports C++14 and newer.
  200 | static_assert(PROTOBUF_CPLUSPLUS_MIN(201402L), "Protobuf only supports C++14 and newer.");
      |               ^~~~~~~~~~~~~~~~~~~~~~
/u/yuriv/local/include/google/protobuf/port_def.inc:200:15: note: the comparison reduces to ‘(201103 >= 201402)’
In file included from /u/yuriv/local/include/google/protobuf/stubs/common.h:44,
                 from /u/yuriv/local/include/google/protobuf/io/coded_stream.h:130,
                 from /u/yuriv/nn-trainer/deps/onnx/onnx/onnx-ml.pb.h:24:
```

### System information
- OS Platform and Distribution: Linux (3.10.0-1160.31.1.el7.x86_64)
- ONNX version (*e.g. 1.13*): current master
- Python version: 3.9
- GCC/Compiler version (if compiling from source): gcc-12.2.0
- CMake version: 3.22.3
- Protobuf version: protobuf's master rev. 626c7e7734662b7017402a0be137e2409c828b00

**Comment by Niklesh99** [comment]
Hi @yurivict,

Yes, upgrading the version from 11 to 14 in CMakeList.txt helped for me.
Thanks!

**Comment by jcwchen** [comment]
Thanks for the update! I will upgrade the required CMAKE_CXX_STANDARD version by this PR: https://github.com/onnx/onnx/pull/5119.

**Comment by snnn** [comment]
It means people who are still using CentOS 7.x need to use a compiler from Red Hat Devtoolset, the default one came with the system will not work. Starting from the upcoming ONNX release, ONNX will require gcc version >=5.

**Comment by jcwchen** [comment]
@snnn Thanks for the reminder! IIRC, some of ONNX related projects (for example, openvino) might still use CentOS 7 in CI. cc @p-wysocki for awareness. Note that this CMAKE_CXX_STANDARD upgrade is only included in the main branch, not upcoming 1.14.0.
